src/util/worker/mod.rs

```diff
@@ -22,94 +22,264 @@ quick_error! {
     }
 }
 
+impl<T> From<mpsc::SendError<T>> for Error {
+    fn from(_: mpsc::SendError<T>) -> Error {
+        // do we need to return the failed data
+        Error::Stopped
+    }
+}
+
 pub type Result<T> = result::Result<T, Error>;
 
 pub trait Runnable<T: Display> {
     fn run(&mut self, t: T);
 }
 
+pub trait BatchRunnable<T: Display> {
+    /// run a batch of tasks.
+    ///
+    /// Please note that ts will be clear after invoking this method.
+    fn run_batch(&mut self, ts: &mut Vec<T>);
```

Comment:
No, `Drain` is not as flexible as `Vec`. For example it will be impossible to iterate `Drain` twice.